osu.Game/Screens/Play/GameplayClockContainer.cs

```diff
@@ -296,5 +303,23 @@ public HardwareCorrectionOffsetClock(IClock source, bool processSource = true)
             {
             }
         }
+
+        # region Storyboard outro logic
+
+        public IBindable<bool> HasStoryboardEnded => hasStoryboardEnded;
+
+        public bool HasTimeLeftInStoryboard => GameplayClock.CurrentTime <= StoryboardEndTime;
+
+        private readonly BindableBool hasStoryboardEnded = new BindableBool(true);
+
+        private void updateHasStoryboardEnded()
+        {
+            if (StoryboardEndTime == 0)
+                return;
+
+            hasStoryboardEnded.Value = GameplayClock.CurrentTime >= StoryboardEndTime;
+        }
+
+        # endregion
```

Comment:
Better suited in `GameplayBeatmap` instead.

Reply:
I'm not sure I understand why. `GameplayBeatmap` doesn't have any reference to a storyboard, but `WorkingBeatmap` does. What is the difference between the two?

Reply:
`GameplayClockContainer` shouldn't really have this logic in there, looks to be irrelevant from what it's doing. 

`GameplayBeatmap` is a component cached for DI made to provide information on the beatmap as well as tracking the current gameplay status, I see this logic better suited in that component, you can pass the storyboard along the beatmap there in the constructor. (FWIW, `BreakTracker`, a component that tracks whether the gameplay is in break time, should exist in `GameplayBeatmap` instead).

But now that you mention it, this logic could actually be in `DrawableStoryboard` instead, expose it from `DimmableStoryboard` for `Player` to access and there you have it.

Reply:
I see. Though, why define the logic in `DrawableStoryboard` instead of `DimmableStoryboard`? Additionally, how should I keep track of the time in `DrawableStoryboard`? I tried passing in the `GameplayClock` from `Player`, but then it gets messy since I have to change the constructor and `Storyboard.CreateDrawable()`. Is there a better approach?

Reply:
`DimmableStoryboard` should do nothing but hiding, dimming, or blurring the storyboard, `DrawableStoryboard` is where this sort of logic should go into. (I also envision `LatestEndTime` to be moved there as well rather than in `Storyboard`, as elements like sample and video can't calculate their length without actually accessing the file and reading from there, but that's for another day)

As for the difficulty in passing `GameplayClock`, that's why [dependency injection](https://github.com/ppy/osu-framework/wiki/Dependency-Injection) is a thing, in fact, `DrawableStoryboard` already resolves `GameplayClock` and respects it:
https://github.com/ppy/osu/blob/215ad780454bb0f9fd2a238e1b3334e7a0ee0c61/osu.Game/Storyboards/Drawables/DrawableStoryboard.cs#L62-L66

So you can keep track of time by [`Time.Current`](https://github.com/ppy/osu-framework/blob/99786238a02e0d6b69da86dd52e5506ee8ec0566/osu.Framework/Graphics/Transforms/Transformable.cs#L25-L28), which is set to `GameplayClock` in here.